### ISSUE TITLE
More Feedback

### ISSUE THREAD

**Issue by GaelGil** [GaelGil/algorithms#4]
In `spam_analysis` I have a file called words `words.py`. In there I have a bunch of functions that check if a message is spam or not using individual words. It fails of course but once those functions are done I added another function called `right_or_wrong.py`. In this function I compare the result my function returned with the real label of the message. If the function was correct we print out correct. If it fails I print out fail and append that message to the dictionary. 

Once that's all done I run a function called `check_again.py`which runs everything again and all the previous fails are now correct. So my question is am I allowed to even when I get it wrong just take in the answer because that's what I'm doing in `right_or_wrong.py` function. Even when I'm wrong I take the message and give myself the answer somewhat so technically I could just have a function that looks like this.
~~~python
def check(sms, label):
        return ('ham')

def right_or_wrong(sms:list, result:str, label:str):
    """
    This function takes in the message, result of my function,
    and real label of the function. If the result matches the 
    label everything is cool. If the result does not match the
    """
    if result == label:
        print('correct')
    elif result != label:
        print('fail')
        add_to_spam(sms)

~~~

This would technically give me the same results as my current functions since my function is always returning `ham` and I'm correcting it later anyways. 
I was also reading in some books about supervised ML and this would be like it technically since were using our data that also comes with its labels but am I allowed to use the labels. I also read about how I should look for patterns of spam because writing rules would not work anymore if people changed the way the wrote their spam. 

My main two questions are am I allowed to use the labels and should I change over to using the bigrams rather then words.

**Comment by GaelGil** [comment]
Outputs can be found in `spam_analysis` in the `data_eval` notebook at the bottom

**Comment by camoverride** [comment]
> am I allowed to even when I get it wrong just take in the answer
Yes, you can memorize answers. However, doing this alone won't help you predict sms messages that aren't in the training data: you also wrote "writing rules would not work anymore if people changed the way the wrote their spam", which is basically correct: hand-writing rules might get some results but probably won't scale.

> reading in some books about supervised ML
Yes, this is a supervised ML problem.

> am I allowed to use the labels
You can use the labels to memorize certain sms messages, but it won't let you generalize past the training data. Remember, when I test your algorithm I'll be using data you've never seen before.

> should I change over to using the bigrams rather then words
I actually don't know the answer to this. Machine Learning is experimental, which means you'll have to try both and see what works!

Consider doing something like this, using a pandas dataframe, where `0` is ham and `1` is spam.
:

```
| true_label | predicted_label | message                                 |
| 0               | 0                         | hey are you free tomorrow ? |
| 1               | 1                          | get FREE stuff today!             |
| 0               | 0                         | I'm angry at you :(                  |
| 1               | 0                         | visit www.freestuff.com .       |
| 0               | 0                         | hey are you free ?                  |
| 0               | 0                         | sup dude                                |
```
The `predicted_label` column will get added by your algorithm.

Consider the different types of errors and how to count them:
- type 1: the true label is `0` but you predict `1`.
- type 2: the true label is `1` but you predict `0`.

https://en.wikipedia.org/wiki/Type_I_and_type_II_errors

Maybe you should report each of these types of errors differently? What's worse, when your algorithm thinks a message is spam, but it's legit? Or when your algorithm thinks an algorithm is legit but is actually spam? Imagine a different scenario when instead of detecting spam you're detecting whether some "rustling in the bushes" is a lion: what's worse, confusing the bushes for a lion (and getting scared) or confusing the lion for some bushes (and getting eaten)?
